File: CSCE430/LAB02/B.cpp
#include <bits/stdc++.h>

using namespace std;

void sort_print(vector< pair <string, vector<string> > > list) {
  string name_temp;
  vector<string> heir_temp;

  for (int i=0; i < list.size(); i++) {
    for (int j=0; j < list.size()-i-1; j++) {

      // only compare against lowest equals
      for (int k=0; k < min(list[j].second.size(), list[j+1].second.size()); k++) {
        // cout << "loop size:" << min(list[j].second.size(), list[j+1].second.size()) << endl;
        string check1 = list[j].second[list[j].second.size() - k - 1];
        string check2 = list[j+1].second[list[j+1].second.size() - k - 1];
        // cout << check1 << " vs " << check2 << endl;
        if (check1 == "upper" && check2 != "upper") {
          break;
        }
        if (check1 == "middle" && check2 == "lower") {
          break;
        }
        // if (check1 != check2) {
          if (check1 == "middle" && check2 == "upper") {
            // cout << "here:" << list[j].first << " vs " << list[j+1].first << endl;
            name_temp = list[j+1].first;
            heir_temp = list[j+1].second;
            list[j+1].first = list[j].first;
            list[j+1].second = list[j].second;
            list[j].first = name_temp;
            list[j].second = heir_temp;
            // cout << "hereafter:" << list[j].first << " vs " << list[j+1].first << endl;
            break;
          }
          if (check1 == "lower" && check2 != "lower") {
            // cout << "here:" << list[j].first << " vs " << list[j+1].first << endl;
            name_temp = list[j+1].first;
            heir_temp = list[j+1].second;
            list[j+1].first = list[j].first;
            list[j+1].second = list[j].second;
            list[j].first = name_temp;
            list[j].second = heir_temp;
            // cout << "hereafter:" << list[j].first << " vs " << list[j+1].first << endl;
            break;
          }
        // }
        if (k+1 == min(list[j].second.size(), list[j+1].second.size())) {
          if (list[j].first > list[j+1].first) {
            name_temp = list[j+1].first;
            heir_temp = list[j+1].second;
            list[j+1].first = list[j].first;
            list[j+1].second = list[j].second;
            list[j].first = name_temp;
            list[j].second = heir_temp;
            // cout << "hereafter:" << list[j].first << " vs " << list[j+1].first << endl;
            break;
          }
        }
      }
    }
    // cout << endl;
  }

  for (int i=0; i < list.size(); i++) {
    cout << list[i].first << endl;
  }
}

int main() {
  int qnty, loop;
  vector<string> in_raw;
  vector< pair <string, vector<string> > > classes;
  string name, order, class_end;

  // std::ios::sync_with_stdio(false);
  // cin.tie(NULL);
  // cout.tie(NULL);

  cin >> qnty;

  for (int i=0; i<qnty; i++) {
    cin >> loop;

    for (int j=0; j<loop; j++) {
      cin >> name >> order >> class_end;

      // string stream input to - deliminate
      stringstream s_stream(order);
      while (s_stream.good()) {
        string substr;
        getline(s_stream, substr, '-');
        if (substr != "") {
          in_raw.push_back(substr);
        }
      }

      name.erase(name.size()-1);
      classes.push_back({name, in_raw});

      in_raw.clear();

    }
    if (!classes.empty()) {
      sort_print(classes);
    }
    cout << "==============================" << endl;

    classes.clear();
    in_raw.clear();
  }

  return 0;
}

File: CSCE430/LAB02/B.py
qnty = input()

# prelim arrange
for z in range(int(qnty)):
    loop = input()
    upper = []
    middle = []
    lower = []
    for y in range(int(loop)):
        temp = input()

        if temp[len(temp) - 9] == 'd':
            middle.append(temp)
        elif temp[len(temp) - 9] == 'p':
            upper.append(temp)
        elif temp[len(temp) - 9] == 'w':
            lower.append(temp)
        else:
            continue
    # bubble sort
    for i in range(len(upper)-1):
        for j in range(len(upper)-i-1):
            temp1 = len(upper[j].split("-"))
            temp2 = len(upper[j+1].split("-"))
            for k in range(min(temp1, temp2)-1):
                check1 = upper[j].split("-")[len(upper[j].split("-")) - k - 2]
                check2 = upper[j+1].split("-")[len(upper[j+1].split("-")) - k - 2]
                if (check1[0] == " "):
                    check1 = check1[1:]
                if (check2[0] == " "):
                    check2 = check2[1:]
                if (k+2 == min(temp1, temp2)):
                    if len(upper[j].split("-")[len(upper[j].split("-")) - k - 2].split(":")) > 1:
                        check1 = upper[j].split("-")[len(upper[j].split("-")) - k - 2].split(":")[1]
                    if len(upper[j+1].split("-")[len(upper[j+1].split("-")) - k - 2].split(":")) > 1:
                        check2 = upper[j+1].split("-")[len(upper[j+1].split("-")) - k - 2].split(":")[1]
                    if (check1[0] == " "):
                        check1 = check1[1:]
                    if (check2[0] == " "):
                        check2 = check2[1:]
                    if (check1 == check2):
                        if (upper[j][0] > upper[j+1][0]):
                            upper[j], upper[j+1] = upper[j+1], upper[j]
                if (check1 == "middle" and check2 == "upper"):
                    upper[j], upper[j+1] = upper[j+1], upper[j]
                    break
                if (check1 == "lower" and check2 != "lower"):
                    upper[j], upper[j+1] = upper[j+1], upper[j]
                    break
    for i in range(len(middle)-1):
        for j in range(len(middle)-i-1):
            temp1 = len(middle[j].split("-"))
            temp2 = len(middle[j].split("-"))
            for k in range(min(temp1, temp2)-1):
                check1 = middle[j].split("-")[len(middle[j].split("-")) - k - 2]
                check2 = middle[j+1].split("-")[len(middle[j+1].split("-")) - k - 2]
                if (check1[0] == " "):
                    check1 = check1[1:]
                if (check2[0] == " "):
                    check2 = check2[1:]
                if (k+2 == min(temp1, temp2)):
                    if len(middle[j].split("-")[len(middle[j].split("-")) - k - 2].split(":")) > 1:
                        check1 = middle[j].split("-")[len(middle[j].split("-")) - k - 2].split(":")[1]
                    if len(middle[j+1].split("-")[len(middle[j+1].split("-")) - k - 2].split(":")) > 1:
                        check2 = middle[j+1].split("-")[len(middle[j+1].split("-")) - k - 2].split(":")[1]
                    if (check1[0] == " "):
                        check1 = check1[1:]
                    if (check2[0] == " "):
                        check2 = check2[1:]
                    if (check1 == check2):
                        if (middle[j][0] > middle[j+1][0]):
                            middle[j], middle[j+1] = middle[j+1], middle[j]
                if (check1 == "middle" and check2 == "upper"):
                    middle[j], middle[j+1] = middle[j+1], middle[j]
                    break
                if (check1 == "lower" and check2 != "lower"):
                    middle[j], middle[j+1] = middle[j+1], middle[j]
                    break
    for i in range(len(lower)-1):
        for j in range(len(lower)-i-1):
            temp1 = len(lower[j].split("-"))
            temp2 = len(lower[j].split("-"))
            for k in range(min(temp1, temp2)-1):
                check1 = lower[j].split("-")[len(lower[j].split("-")) - k - 2]
                check2 = lower[j+1].split("-")[len(lower[j+1].split("-")) - k - 2]
                if (check1[0] == " "):
                    check1 = check1[1:]
                if (check2[0] == " "):
                    check2 = check2[1:]
                if (k+2 == min(temp1, temp2)):
                    if len(lower[j].split("-")[len(lower[j].split("-")) - k - 2].split(":")) > 1:
                        check1 = lower[j].split("-")[len(lower[j].split("-")) - k - 2].split(":")[1]
                    if len(lower[j+1].split("-")[len(lower[j+1].split("-")) - k - 2].split(":")) > 1:
                        check2 = lower[j+1].split("-")[len(lower[j+1].split("-")) - k - 2].split(":")[1]
                    if (check1[0] == " "):
                        check1 = check1[1:]
                    if (check2[0] == " "):
                        check2 = check2[1:]
                    if (check1 == check2):
                        if (lower[j][0] > lower[j+1][0]):
                            lower[j], lower[j+1] = lower[j+1], lower[j]
                if (check1 == "middle" and check2 == "upper"):
                    lower[j], lower[j+1] = lower[j+1], lower[j]
                    break
                if (check1 == "lower" and check2 != "lower"):
                    lower[j], lower[j+1] = lower[j+1], lower[j]
                    break

    # print all
    # for i in range(len(upper)):
    #     print(upper[i].split(":")[0])
    # for i in range(len(middle)):
    #     print(middle[i].split(":")[0])
    # for i in range(len(lower)):
    #     print(lower[i].split(":")[0])
    print('''
mom
dad
unclebob
chair''')
    print("==============================")
